File: js/endless-binary/dec2bin-2.js
// taidalab Version 1.4.0
// https://github.com/taidalog/taidalab
// Copyright (c) 2022 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
function checkAnswerd2b2 (answer, last_answers) {
    // Getting the user input.
    const numberInput = document.getElementById('numberInput');
    const bin = escapeHtml(numberInput.value);
    console.log(bin);
    
    numberInput.focus();
    
    // Making an error message.
    const errorMessage = newErrorMessageBin(answer, bin);
    document.getElementById('errorArea').innerHTML = errorMessage;
    
    // Exits when the input was invalid.
    if (errorMessage) {
        return;
    }
    
    // Converting the input in order to use in the history message.
    const binaryDigit = 8;
    const destinationRadix = 2;
    const zeroPaddedBin = bin.padStart(binaryDigit, '0');
    const taggedBin = colorLeadingZero(zeroPaddedBin);
    const dec = parseInt(bin, destinationRadix);
    console.log(taggedBin);
    console.log(dec);
    
    const decimalDigit = 3;
    const spacePaddedDec = dec.toString().padStart(decimalDigit, ' ').replace(' ', '&nbsp;');
    
    // Making a new history and updating the history with the new one.
    const sourceRadix = 10;
    const outputArea = document.getElementById('outputArea');
    const currentHistoryMessage = newHistory((dec == answer), taggedBin, destinationRadix, spacePaddedDec, sourceRadix);
    const historyMessage = concatinateStrings(currentHistoryMessage, outputArea.innerHTML);
    console.log(currentHistoryMessage);
    console.log(historyMessage);
    outputArea.innerHTML = historyMessage;
    
    if (dec == answer) {
        // Making the next question.
        let nextNumber = 0;

        console.log(last_answers);
        do {
            nextNumber = getRandomBetween(0, 255);
            console.log(nextNumber);
            console.log(last_answers.some((element) => element == nextNumber));
        } while (last_answers.some((element) => element == nextNumber));

        document.getElementById('questionSpan').innerText = nextNumber;
        numberInput.value = '';

        // Updating `lastAnswers`.
        // These numbers will not be used for the next question.
        const answersToKeep = 10;
        const lastAnswers = [nextNumber].concat(last_answers).slice(0, answersToKeep);

        // Setting the next answer to the check button.
        document.getElementById('submitButton').onclick = function() { checkAnswerd2b2(nextNumber, lastAnswers); return false; };
    }
}


function initDec2Bin2 () {
    // Initialization.
    const initNumber = getRandomBetween(0, 255);
    const sourceRadix = 10;
    const destinationRadix = 2;
    
    document.getElementById('questionSpan').innerHTML = initNumber;
    document.getElementById('srcRadix').innerHTML = '(' + sourceRadix + ')';
    document.getElementById('dstRadix').innerHTML = destinationRadix;
    document.getElementById('binaryRadix').innerHTML = '<sub>(' + destinationRadix + ')</sub>';
    document.getElementById('submitButton').onclick = function() { checkAnswerd2b2(initNumber, [initNumber]); return false; };
}

File: src/Dec2Bin2.fs
﻿// taidalab Version 1.6.0
// https://github.com/taidalog/taidalab
// Copyright (c) 2022 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
namespace Taidalab

open System
open Browser.Dom
open Taidalab.Common

module Dec2Bin2 =

    let rec checkAnswer answer (last_answers : int list) =
        // Getting the user input.
        let numberInput = document.getElementById "numberInput" :?> Browser.Types.HTMLInputElement
        let bin = escapeHtml numberInput.value
        printfn "bin: %s" bin
        
        numberInput.focus()
        
        // Making an error message.
        let errorMessage = newErrorMessageBin (string answer) bin
        (document.getElementById "errorArea").innerHTML <- errorMessage
        
        // Exits when the input was invalid.
        if errorMessage <> "" then
            ()
        else
            
            // Converting the input in order to use in the history message.
            let binaryDigit = 8
            let destinationRadix = 2
            let taggedBin = bin |> padWithZero binaryDigit |> colorLeadingZero
            let dec = toDecimal bin
            printfn "taggedBin: %s" taggedBin
            printfn "dec: %d" dec
            
            let decimalDigit = 3
            let spacePaddedDec = dec |> string |> padStart " " decimalDigit |> escapeSpace
            
            // Making a new history and updating the history with the new one.
            let sourceRadix = 10
            let outputArea = document.getElementById "outputArea"
            let historyMessage =
                newHistory (dec = answer) taggedBin destinationRadix spacePaddedDec sourceRadix
                |> (fun x-> concatinateStrings "<br>" x outputArea.innerHTML)
            // printfn "historyMessage: %s" historyMessage
            outputArea.innerHTML <- historyMessage
            
            if dec = answer then
                // Making the next question.
                let mutable nextNumber = getRandomBetween 0 255

                printfn "last_answers: %A" last_answers
                while List.contains nextNumber last_answers do
                    nextNumber <- getRandomBetween 0 255
                    printfn "nextNumber: %d" nextNumber
                    printfn "List.contains nextNumber last_answers: %b" (List.contains nextNumber last_answers)

                (document.getElementById "questionSpan").innerText <- string nextNumber
                numberInput.value <- ""

                // Updating `lastAnswers`.
                // These numbers will not be used for the next question.
                let answersToKeep = Math.Min(10, List.length last_answers + 1)
                let lastAnswers = (nextNumber :: last_answers).[0..(answersToKeep - 1)]

                // Setting the next answer to the check button.
                (document.getElementById "submitButton").onclick <- (fun _ -> checkAnswer nextNumber lastAnswers)


    let init () =
        // Initialization.
        let initNumber = getRandomBetween 0 255
        let sourceRadix = 10
        let destinationRadix = 2
        
        (document.getElementById "questionSpan").innerHTML <- string initNumber
        (document.getElementById "srcRadix").innerHTML <- sprintf "(%d)" sourceRadix
        (document.getElementById "dstRadix").innerHTML <- string destinationRadix
        (document.getElementById "binaryRadix").innerHTML <- sprintf "<sub>(%d)</sub>" destinationRadix
        (document.getElementById "submitButton").onclick <- (fun _ -> checkAnswer initNumber [initNumber])